// ravendb/dist/Documents/Operations/Etl/ConnectionString.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueueConnectionString = exports.Authentication = exports.ElasticSearchConnectionString = exports.OlapConnectionString = exports.SqlConnectionString = exports.RavenConnectionString = exports.ConnectionString = void 0;
class ConnectionString {
}
exports.ConnectionString = ConnectionString;
class RavenConnectionString extends ConnectionString {
    constructor() {
        super(...arguments);
        this.type = "Raven";
    }
}
exports.RavenConnectionString = RavenConnectionString;
class SqlConnectionString extends ConnectionString {
    constructor() {
        super(...arguments);
        this.type = "Sql";
    }
}
exports.SqlConnectionString = SqlConnectionString;
class OlapConnectionString extends ConnectionString {
    constructor() {
        super(...arguments);
        this.type = "Olap";
    }
}
exports.OlapConnectionString = OlapConnectionString;
class ElasticSearchConnectionString extends ConnectionString {
    constructor() {
        super(...arguments);
        this.type = "ElasticSearch";
    }
}
exports.ElasticSearchConnectionString = ElasticSearchConnectionString;
class Authentication {
}
exports.Authentication = Authentication;
class QueueConnectionString extends ConnectionString {
    constructor() {
        super(...arguments);
        this.type = "Queue";
    }
}
exports.QueueConnectionString = QueueConnectionString;
